src/bt.rs

```diff
@@ -74,324 +73,217 @@ pub enum BinaryTreeError<V> {
-    /// - `Err(InsertNonEmptyNode(value))` when the tree already contains a
-    /// value at the end of the path.
-    /// - `Err(UnreachablePath(value))` when the end of the path is unreachable.
-    ///
-    /// In the error cases, no insertion occurs and the value is returned to
-    /// the caller of this function.
-    pub fn insert(&mut self, path: &Path, value: V) -> Result<(), BinaryTreeError<V>> {
-        if let Some(node) = &mut self.root {
-            node.insert(path, value)
-        } else if path.is_empty() {
-            self.root = Some(Box::new(Node::new(value)));
-            Ok(())
-        } else {
-            Err(BinaryTreeError::UnreachableNode(value))
+            let next = &self.nodes[node?];
+            node = if !bit { next.left } else { next.right };
         }
-    }
-
-    /// Gets a reference to the value located at the end of the path.
-    ///
-    /// This function traverses the tree from the root node until reaching the
-    /// node at the end of the path. It returns [None], if the node is
-    /// unreachable or nonexistent. Otherwise, it returns a reference to the
-    /// value stored in the node.
-    pub fn get(&self, path: &Path) -> Option<&V> {
-        self.root.as_ref().and_then(|node| node.get(path))
-    }
 
-    /// Gets a mutable reference to the node located at the end of the path.
-    ///
-    /// This function traverses the tree from the root node until reaching the
-    /// node at the end of the path. It returns [None], if the node is
-    /// unreachable or nonexistent. Otherwise, it returns a mutable reference
-    /// to the node.
-    pub fn get_node(&mut self, path: &Path) -> Option<&mut Node<V>> {
-        self.root.as_mut().and_then(|node| node.get_node(path))
+        Some(&mut self.nodes[node?])
     }
 }
 
 impl<V> Default for BinaryTree<V> {
     fn default() -> Self {
         Self {
             root: Option::default(),
+            nodes: Vec::with_capacity(Self::NODES_CAPACITY),
         }
     }
 }
 
-impl<V: Encode> Encode for Node<V> {
-    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
-        self.value.encode(bytes)?;
+// Indicates the datatype used to serialize usize values.
+type UsizeType = u16;
 
-        let mut stack = Vec::new();
-        stack.push(&self.right);
-        stack.push(&self.left);
-
-        // Nodes are stored following a pre-order traversal.
-        while let Some(elem) = stack.pop() {
-            match elem {
-                None => CodecMarker::Leaf.encode(bytes)?,
-                Some(node) => {
-                    CodecMarker::Inner.encode(bytes)?;
-                    node.value.encode(bytes)?;
-                    stack.push(&node.right);
-                    stack.push(&node.left);
-                }
-            }
+impl Encode for Option<usize> {
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        match *self {
+            None => UsizeType::encode(&0, bytes),
```

Comment:
While this doesn't cause functional issues due to the way higher-level encoding/decoding functions are written, IMO it's questionable to encode `None` as a zero value; specifically, `None` and `Some(0)` are encoded in the same way, and `Some(0)` becomes `None` when roundtripped through encoding.

(This doesn't cause problems because, in practice, we never decode `Some(0)` for an encoding of a valid tree: decoding of `Option<usize>` is done only for the indices of child nodes, and the root is always at index 0, so index 0 will never be a valid child index.)

I think this is especially important to address if we decide to keep this as an encoding on `Option<usize>` rather than `Option<SomeNewtype>`, since in that case anyone using `Encode` will be using this implementation & they are unlikely to expect this behavior.